components/html-block/html-block.js

```diff
@@ -0,0 +1,270 @@
+				const HTMLHandler = window.MathJax._.handlers.html.HTMLHandler.HTMLHandler;
+				const AbstractHandler = window.MathJax._.core.Handler.AbstractHandler.prototype;
+				const startup = window.MathJax.startup;
+
+				//
+				//  Extend HTMLAdaptor to handle shadowDOM as the document
+				//
+				class ShadowAdaptor extends HTMLAdaptor {
+					body(doc) {
+						return doc.body || (doc.firstChild || {}).lastChild || doc;
+					}
+					create(kind, ns) {
+						const document = (this.document.createElement ? this.document : this.window.document);
+						return (ns ?
+							document.createElementNS(ns, kind) :
+							document.createElement(kind));
+					}
+					head(doc) {
+						return doc.head || (doc.firstChild || {}).firstChild || doc;
+					}
+					root(doc) {
+						return doc.documentElement || doc.firstChild || doc;
+					}
+					text(text) {
+						const document = (this.document.createTextNode ? this.document : this.window.document);
+						return document.createTextNode(text);
+					}
+				}
+
+				//
+				//  Extend HTMLHandler to handle shadowDOM as document
+				//
+				class ShadowHandler extends HTMLHandler {
+					create(document, options) {
+						const adaptor = this.adaptor;
+						if (typeof(document) === 'string') {
+							document = adaptor.parse(document, 'text/html');
+						} else if ((document instanceof adaptor.window.HTMLElement || document instanceof adaptor.window.DocumentFragment) && !(document instanceof window.ShadowRoot)) {
+							const child = document;
+							document = adaptor.parse('', 'text/html');
+							adaptor.append(adaptor.body(document), child);
+						}
+						//
+						//  We can't use super.create() here, since that doesn't
+						//  handle shadowDOM correctly, so call HTMLHandler's parent class
+						//  directly instead.
+						//
+						return AbstractHandler.create.call(this, document, options);
+					}
+				}
+
+				//
+				//  Register the new handler and adaptor
+				//
+				startup.registerConstructor('HTMLHandler', ShadowHandler);
+				startup.registerConstructor('browserAdaptor', () => new ShadowAdaptor(window));
+
+				//
+				//  A service function that creates a new MathDocument from the
+				//  shadow root with the configured input and output jax, and then
+				//  renders the document.  The MathDocument is returned in case
+				//  you need to rerender the shadowRoot later.
+				//
+				window.MathJax.typesetShadow = function(root) {
+					const InputJax = startup.getInputJax();
+					const OutputJax = startup.getOutputJax();
+					const html = mathjax.document(root, { InputJax, OutputJax });
+					html.render();
+					return html;
+				};
+
+				//
+				//  Now do the usual startup now that the extensions are in place
+				//
+				window.MathJax.startup.defaultReady();
+			}
+		}
+	};
+
+	mathJaxLoaded = new Promise(resolve => {
+		const script = document.createElement('script');
+		script.async = 'async';
```

Comment:
I think `async` is correct... it's going to not render-block while downloading, but then it will. `defer` on the other hand would still download asynchronously but wait until HTML parsing is complete before executing. I don't actually think it matters since most web components are going to upgrade _after_ HTML parsing anyway I think.

Reply:
Yeah, I was basically following the config guidance from MathJax here - they were recommending `async` unless the config and the library were both being requested, in which case they recommend `defer` since `async` doesn't guarantee order, and the config is required to come before the library in order for initialization to be correct.  That all being said, I think we can switch it to `defer` if it matters because our config is within this module.  In this particular case it's being requested pretty late, only after we've detected `math` in the DOM.